### PR TITLE
Update AssemblyScript to 0.19.2

### PR DESCRIPTION
**Pull request by otaviopace**
This PR allows subgraph developers to use AssemblyScript on version `0.19.2`. To enable the feature, they must change their `subgraph.yaml`'s `apiVersion` to a version greater than or equal to `0.0.5`. The behavior for the current deployed subgraphs shouldn't change at all.

Before this is merged, we need to release a new version of both `graph-ts` and `graph-cli`, so we can point to them in the `integration-tests`, for now they're using these PRs branches:

- https://github.com/graphprotocol/graph-cli/pull/690
- https://github.com/graphprotocol/graph-ts/pull/185

Important changes:

- runtime/asc_abi/class now holds mappings for both versions, the ones that needed to change were: ArrayBuffer, TypedArray, AscString and Array;
- Some runtime if/elses have been added to support both versions (in WasmInstance/WasmInstanceContext for memory alignment and new exported functions);
- runtime unit tests have been duplicated
- A new integration-test has been added pointing to the older apiVersion 0.0.4

### CODE REVIEW

**Comment by arthas168** [comment]
Awesome job! 🎉

File: runtime/wasm/src/asc_abi/v0_0_5.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+            ty: PhantomData,
+        })
+    }
+
+    pub(crate) fn to_vec<H: AscHeap + ?Sized>(
+        &self,
+        heap: &H,
+    ) -> Result<Vec<T>, DeterministicHostError> {
+        self.buffer.read_ptr(heap)?.get(
+            self.data_start - self.buffer.wasm_ptr(), // needed to get offset
```
Comment (leoyvens) [inline runtime/wasm/src/asc_abi/v0_0_5.rs:133]: Better comment and use checked subtraction (same for `Array<T>`)

**Review by leoyvens — CHANGES_REQUESTED** [review]
Amazing work! I have a few small comments, and we should try to reduce duplication in the unit tests.

File: runtime/wasm/src/asc_abi/v0_0_5.rs (hunk 1 above)
Comment (otaviopace) [reply runtime/wasm/src/asc_abi/v0_0_5.rs]: Solved!

File: runtime/wasm/src/asc_abi/class.rs [hunk 2]
```diff
@@ -1,312 +1,384 @@
...
-        })
+        v0_0_4::AscString::asc_size(AscPtr::new(ptr.wasm_ptr()), heap)
+    }
+
+    fn content_len(&self, asc_bytes: &[u8]) -> usize {
+        if let Self::ApiVersion0_0_5(s) = self {
+            return s.content_len(asc_bytes);
+        }
+
+        unreachable!("Only called for apiVersion >=0.0.5");
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply runtime/wasm/src/asc_abi/class.rs]: Solved!

File: runtime/wasm/src/asc_abi/class.rs [hunk 3]
```diff
@@ -1,312 +1,384 @@
...
-        Ok(byte_length_size + padding_size + byte_length)
+        v0_0_4::ArrayBuffer::asc_size(AscPtr::new(ptr.wasm_ptr()), heap)
+    }
+
+    fn content_len(&self, asc_bytes: &[u8]) -> usize {
+        if let Self::ApiVersion0_0_5(a) = self {
+            return a.content_len(asc_bytes);
+        }
+
+        unreachable!("Only called for apiVersion >=0.0.5");
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply runtime/wasm/src/asc_abi/class.rs]: Solved!

File: runtime/test/src/test.rs [hunk 4]
```diff
@@ -89,11 +99,19 @@ fn test_valid_module_and_store_with_timeout(
...
+) -> WasmInstance<Chain> {
+    test_valid_module_and_store(subgraph_id, data_source, api_version).0
 }
 
-fn mock_data_source(path: &str) -> DataSource {
+fn mock_data_source(wasm_file: &str, api_version: Version) -> DataSource {
+    let path = format!(
+        "wasm_test/api_version_{}_{}_{}/{}",
+        api_version.major, api_version.minor, api_version.patch, wasm_file
+    );
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply runtime/test/src/test.rs:124]: Thanks 😊

File: graph/src/runtime/mod.rs [hunk 5]
```diff
@@ -107,6 +132,75 @@ macro_rules! impl_asc_type {
...
+    Uint64Array = 9,
+    Float32Array = 10,
+    Float64Array = 11,
+    BigDecimal = 12,
+    ArrayBool = 13,
+    ArrayUint8Array = 14,
+    ArrayEthereumValue = 15,
+    ArrayStoreValue = 16,
+    ArrayJsonValue = 17,
+    ArrayString = 18, // *
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply graph/src/runtime/mod.rs]: Removed!

File: graph/src/runtime/asc_ptr.rs [hunk 6]
```diff
@@ -71,6 +103,58 @@ impl<C: AscType> AscPtr<C> {
         Ok(u32::from_le_bytes(u32_bytes))
     }
 
+    /// Helper that generates an AssemblyScript (v0.19.2) header.
+    /// Only used for version >= 0.0.5.
+    fn generate_header<H: AscHeap + ?Sized>(
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply graph/src/runtime/asc_ptr.rs:116]: Solved!

File: graph/src/runtime/asc_ptr.rs [hunk 7]
```diff
@@ -71,6 +103,58 @@ impl<C: AscType> AscPtr<C> {
         Ok(u32::from_le_bytes(u32_bytes))
     }
 
+    /// Helper that generates an AssemblyScript (v0.19.2) header.
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply graph/src/runtime/asc_ptr.rs]: Removed!

File: graph/src/runtime/asc_ptr.rs [hunk 8]
```diff
@@ -49,20 +50,51 @@ impl<C: AscType> AscPtr<C> {
...
+        match heap.api_version() {
+            version if version <= Version::new(0, 0, 4) => {
+                let heap_ptr = heap.raw_new(&asc_obj.to_asc_bytes()?)?;
+                Ok(AscPtr::new(heap_ptr))
+            }
+            _ => {
+                let mut bytes = asc_obj.to_asc_bytes()?;
+
+                let aligned_len = get_aligned_length(bytes.len());
+                bytes.extend(std::iter::repeat(0).take(aligned_len));
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply graph/src/runtime/asc_ptr.rs:80]: Solved!

File: graph/src/runtime/mod.rs [hunk 9]
```diff
@@ -141,3 +235,9 @@ impl From<DeterministicHostError> for HostExportError {
         HostExportError::Deterministic(value.0)
     }
 }
+
+pub const HEADER_SIZE: usize = 20;
+
+pub fn get_aligned_length(content_length: usize) -> usize {
+    16 - (HEADER_SIZE + content_length) % 16
+}
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply graph/src/runtime/mod.rs:243]: Solved!

File: runtime/derive/src/lib.rs [hunk 10]
```diff
@@ -77,18 +78,32 @@ fn asc_type_derive_struct(item_struct: ItemStruct) -> TokenStream {
...
+                // Sanity check
+                match &api_version {
+                    api_version if *api_version <= Version::new(0, 0, 4) => {
+                        if asc_obj.len() < std::mem::size_of::<Self>() {
+                            return Err(DeterministicHostError(anyhow::anyhow!("Size does not match")));
+                        }
+                    }
+                    _ => {
+                        let content_size = size_of::<Self>();
+                        let aligned_size = get_aligned_length(content_size);
```
[the remark this answers is not in the document]
Comment (otaviopace) [reply runtime/derive/src/lib.rs]: Solved!

File: tests/integration-tests/host-exports/src/mapping.ts [hunk 11]
```diff
@@ -1,7 +1,9 @@
 import { Trigger } from "../generated/Contract/Contract";
 import { Address, BigDecimal, BigInt, ethereum } from "@graphprotocol/graph-ts";
 
-// Test that host exports work in globals.
+// TODO: remove this, graph-ts should guarantee this
+__alloc(0);
```
Comment (leoyvens) [inline tests/integration-tests/host-exports/src/mapping.ts:5]: Can this TODO be removed?

Comment (otaviopace) [reply tests/integration-tests/host-exports/src/mapping.ts:5]: Yes, this should be removed 😅 

I'll remove both TODO and `__alloc`.

The `__alloc` should be called by my `graph-ts` branch.

**Review by leoyvens — APPROVED** [review]

**Comment by leoyvens** [comment]
Lets just squash the commits when merging

File: tests/integration-tests/host-exports/src/mapping.ts (hunk 11 above)
Comment (otaviopace) [reply tests/integration-tests/host-exports/src/mapping.ts]: Removed!

File: runtime/derive/src/lib.rs [hunk 12]
```diff
@@ -77,18 +78,38 @@ fn asc_type_derive_struct(item_struct: ItemStruct) -> TokenStream {
...
             #[allow(unused_variables)]
-            fn from_asc_bytes(asc_obj: &[u8]) -> Result<Self, DeterministicHostError> {
-                if asc_obj.len() != std::mem::size_of::<Self>() {
-                    return Err(DeterministicHostError(anyhow::anyhow!("Size does not match")));
-                }
+            fn from_asc_bytes(asc_obj: &[u8], api_version: semver::Version) -> Result<Self, DeterministicHostError> {
+                // Sanity check
+                match &api_version {
+                    api_version if *api_version <= Version::new(0, 0, 4) => {
+                        // This was using an double equal sign before instead of less than.
```
Comment (otaviopace) [inline runtime/derive/src/lib.rs:85]: @leoyvens here's the final comment I've talked about

**Pull request merged**